### PR TITLE
perf(table): support compressed index blocks

### LINKED ISSUE #170: perf(table): index block restart intervals > 1 for prefix truncation
## Context

Upstream: fjall-rs/lsm-tree#184

Currently index blocks use restart interval = 1, meaning every index entry stores the full key. This wastes space when keys share long prefixes — common in graph databases where keys follow patterns like `adj:out:<vertex_id>:<edge_label>:<target>`.

## Problem

With restart interval = 1:
- Every index block entry = full key bytes + block handle (offset + size)
- For 1000 SST data blocks with 64-byte average keys → ~64 KiB of index data
- With prefix truncation (restart interval = 16): ~12-20 KiB (3-5× smaller)

Smaller index blocks mean:
1. **Less memory** for pinned indexes (significant when thousands of SSTs open)
2. **Better cache utilization** — more index entries fit in L2/L3 cache
3. **Faster index scans** — fewer bytes to traverse during binary search

## How it works

Data blocks already use restart intervals for prefix compression. The same technique applies to index blocks:

```
Restart interval = 1 (current):
  Entry 0: full_key_0, handle_0     ← restart point
  Entry 1: full_key_1, handle_1     ← restart point
  Entry 2: full_key_2, handle_2     ← restart point

Restart interval = 16 (proposed):
  Entry 0:  full_key_0, handle_0    ← restart point
  Entry 1:  shared=40, suffix_1, handle_1    ← prefix-compressed
  ...
  Entry 15: shared=38, suffix_15, handle_15  ← prefix-compressed
  Entry 16: full_key_16, handle_16  ← restart point
```

Binary search operates on restart points (every 16th entry), then linear scan within the restart interval. With k=7 hash probes in binary search over restart points, the linear scan adds at most 15 comparisons — negligible vs I/O.

### Additional optimization

Non-restart-head entries can skip encoding the block offset entirely — it can be derived from the previous entry's offset + size. This saves 4-8 bytes per entry (varint encoded offset).

## Implementation plan

1. Add `index_block_restart_interval: u32` config field (default = 1 for backward compat)
2. Extend index block writer to use prefix compression when interval > 1
3. Extend index block reader to handle restart points in seek
4. Add SST metadata field to store the interval (for reader to know how to decode)
5. Benchmark: index block size reduction + point read throughput

## Files

- `src/table/index_block/mod.rs` — reader changes
- `src/table/index_block/iter.rs` — seek with restart intervals
- `src/table/writer/mod.rs` — writer changes
- `src/config/mod.rs` — new config field

## Estimate

**3d** (writer: 1d, reader+seek: 1.5d, tests+bench: 0.5d)

## References

- Upstream: fjall-rs/lsm-tree#184
- RocksDB: `index_block_restart_interval` option (default=1, recommended=16 for large DBs)
- LevelDB: fixed restart interval=16 in both data and index blocks

### LINKED ISSUE #183: fix(index-block): validate shared prefix against restart-key span
## Context

Found during review of PR #182. Consolidated from #185 and #186 (closed as duplicates).

## Problem

`parse_truncated` in `src/table/index_block/block_handle.rs` validates `shared_prefix_len` only against the current entry boundary (`prefix_end <= offset`). It does **not** verify that the shared prefix stays within the actual restart-key span.

With malformed on-disk data (corrupted or crafted block), a `shared_prefix_len` value can be within entry buffer bounds yet exceed the length of the restart-key for that interval. For restart intervals with multiple truncated entries, this allows a shared prefix that extends beyond the restart key bytes — causing silent mis-reconstruction of prefix-compressed keys.

This is a **semantic** validation gap (not a buffer-overflow — strict buffer/offset bounds were already hardened in PR #182). The risk is semantically invalid key reconstruction, not memory safety.

## Proposed approach

1. Carry restart-key span (start + length) through decoder state into the truncated parse path
2. Validate `shared_prefix_len <= restart_key_len` **before** constructing `SliceIndexes`
3. Return corruption error on violation (not panic)

This is a parser-shape change: `parse_truncated` needs access to the restart-key span, which it currently does not receive.

## Scope

- `src/table/index_block/block_handle.rs` — `parse_truncated`: add restart-key span parameter, validate
- `src/table/block/decoder.rs` — thread restart-key span through decoder state
- `src/table/index_block/mod.rs` — wire span into parse calls
- Tests: corruption regression test with malformed `shared_prefix_len` exceeding restart-key length

## Estimate

~4h (implementation 2.5h, tests 1h, review/cleanup 0.5h)

## Links

- PR: #182
- Closed duplicates: #185, #186

### LINKED ISSUE #189: bug(index): seek_upper_bound_cursor returns only first block when all blocks share end_key
## Bug

`IndexBlockIter::seek_upper_impl` with `check_back_cache=false` (forward iteration limit, called via `seek_upper_bound_cursor` → `Table::Iter::next`) uses predicate `end_key < needle` with `partition_point_2`. When **all blocks in an SST share the same `end_key`** (e.g. a key with thousands of merge operands spread across many SST blocks), the predicate returns `false` for every entry, `partition_point_2` returns index 0 (first block), and `hi_scanner.offset` is set to just after the first block handle — so only 1 block is ever served by forward iteration.

## Symptom

`StorageEngine::get(Partition::Adj, key)` after writing 4000 merge operands for the same user key returns only ~121 UIDs (contents of first SST block) instead of 4000.

Failing tests in `coordinode-storage`:
- `merge_concurrent_multithreaded_adds` — expected 4000 UIDs, got 121
- `merge_compaction_preserves_time_travel`
- `merge_stress_concurrent_writers_zero_conflict`
- `merge_stress_interleaved_add_remove_concurrent`
- `merge_time_travel_snapshot_correctness`

## Root cause

In `src/table/index_block/iter.rs`, `seek_upper_impl`:

```rust
// restart_interval == 1 path:
self.decoder.inner_mut().seek_upper(
    |end_key, _s| cmp.compare(end_key, needle) == std::cmp::Ordering::Less,  // < needle
    true,  // partition_point_2
)
```

With pred `end_key < needle` and all `end_key == needle`:
- pred is false for ALL entries
- `partition_point_2` returns index 0
- `fill_stack` fills only the first block handle
- `hi_scanner.offset` = byte offset after first block handle
- `Decoder::next()` stops after first block — **all subsequent blocks are skipped**

## Fix

Split the `restart_interval == 1` path by `check_back_cache`:

- **Back cursor** (`check_back_cache=true`): keep `< needle` — correctly finds first block with `end_key >= needle` for reverse iteration
- **Forward limit** (`check_back_cache=false`): use `<= needle` (`!= Greater`) — `partition_point_2` when all entries satisfy `<=` returns the **last** entry, setting `hi_scanner.offset` after the last block handle → all blocks visited

```rust
if check_back_cache {
    self.decoder.inner_mut().seek_upper(
        |end_key, _s| cmp.compare(end_key, needle) == std::cmp::Ordering::Less,
        true,
    )
} else {
    self.decoder.inner_mut().seek_upper(
        |end_key, _s| cmp.compare(end_key, needle) != std::cmp::Ordering::Greater,
        true,
    )
}
```

## Verification

With the fix: 973/973 lsm-tree tests pass, 174/174 coordinode-storage tests pass (including all 5 previously failing merge tests).

## Note for PR #182

This bug is in `src/table/index_block/iter.rs` which is the same file being modified in PR #182. The fix must be incorporated there.

### LINKED ISSUE #190: perf(decoder): thread entries_end into parse_restart_key() to bound binary-search probes without full-item decode
## Summary

`get_key_at` currently calls `Item::parse_full` on every binary-search probe inside `partition_point` / `partition_point_2`. This decodes the entire restart-head item when only the key slice offsets and `SeqNo` are needed.

## Proposed Change

Thread `entries_end` into `parse_restart_key()` (or a new bounded variant) so that:

- The same restart-head vs. `entries_end` bounds check enforced by `parse_full` is retained, preventing malformed key lengths from leaking into trailer/index bytes.
- Only the minimal data required by the binary-search predicate (key slice and `SeqNo`) is extracted, avoiding the full-item decode cost on each probe.

This would allow `get_key_at` to call the lighter `parse_restart_key` path instead of `parse_full` for `partition_point*` probes.

## Context

Identified during review of PR #182 (compressed index blocks). The bounds-checking introduction in that PR made `get_key_at` switch from unsafe `get_unchecked` + `parse_restart_key` to a safe but heavier `parse_full` call. This issue tracks restoring the lighter path while keeping the new safety invariants.

Backlink: https://github.com/structured-world/coordinode-lsm-tree/pull/182#discussion_r3011292302

Requested by @polaz.

### PR DESCRIPTION
**Pull request by polaz**
## Summary
- enable index block restart intervals > 1 for index block encoding, metadata, and seek paths
- fix seek_upper behavior across compressed restart intervals by advancing to the next restart head when the trimmed tail is still below needle
- fix restart_interval=1 upper-seek limit for all-same-end-key blocks by splitting seek_upper_impl behavior by cursor mode
- harden data_block and index_block parsers: checked_add for all offset arithmetic, reject key/value spans exceeding block boundaries
- harden decoder: reject EOF offsets in reader_at, reject zero-length binary indexes, fail-close forward iteration after back-cursor corruption (poison_back_cursor delegates to clamp_upper_to_lo so next() also stops), clear hi_scanner stack at fill_stack entry
- fix dead is_none() guard in both seek_upper and advance_upper_restart_interval post-fill_stack — use stack.is_empty() to detect corruption after poison_back_cursor switched to clamp semantics
- fail-close seek_upper_impl after clamped advance_upper_restart_interval
- narrow decoder module visibility to pub(crate), narrow peek reset helpers to pub(crate)
- thread entries_end into parse_restart_key for bounded binary-search probes (avoids full-item decode, validates key spans against entries_end)
- propagate meta_partition_size through use_partitioned_filter (was missing, matching use_partitioned_index)
- add pre-write guard to use_partitioned_filter preventing mid-stream filter-writer swap
- fix misleading "compression policy" panic message in RestartIntervalPolicy::new
- add doc comments to encode_into_vec_with_restart_interval, BlockIndexWriter trait method, seek_upper semantics, base_key_end validation
- add regression tests: EOF binary-index offsets, zero-length binary index, poison-back-cursor fail-close, corrupted restart-head key, empty-policy rejection, post-write filter/index swap rejection
- align release automation by disabling publish in release-plz config (git_only=true, publish=false)

## Testing
- cargo fmt --all -- --check
- cargo nextest run --workspace (1017 tests, 0 failed)
- cargo clippy --all-targets -- -D warnings
- cargo test --doc --all-features
- cargo check --all-features (tools/db_bench)
- cargo bench --bench index_block --no-run

## Related
- #184 — make block decoder trailer validation fallible (tracked separately)

Closes #170
Closes #189
Closes #183
Closes #190

## Summary by CodeRabbit

* **New Features**
  * Index encoding now supports configurable restart intervals (rejects zero); writers expose a builder to set it.
  * New public APIs to encode index blocks with custom restart intervals.

* **Bug Fixes**
  * Decoding/iteration now fail-closed on truncated/corrupted blocks, preventing panics and out-of-bounds reads.
  * Configuration validates non-zero/ non-empty restart policies and prevents changing intervals after writing starts.

* **Tests**
  * Added benchmarks and many unit/integration tests for restart intervals, decoding robustness, and point reads.

* **Chores**
  * Release flow set to git-only; crates.io publishing disabled.

### CODE REVIEW

File: src/table/block/decoder.rs [hunk 1]
```diff
@@ -436,15 +501,267 @@ impl<'a, Item: Decodable<Parsed>, Parsed: ParsedItem<Item>> Decoder<'a, Item, Pa
...
+            ) else {
+                break;
+            };
+
+            if !pred(&item, &self.block.data) {
+                break;
+            }
+
+            self.hi_scanner.stack.pop();
+        }
```
[the remark this answers is not in the document]
Comment (polaz) [reply src/table/block/decoder.rs]: @coderabbitai Fixed. `trim_back_while()` now recomputes the forward boundary after popping (`hi_scanner.offset` is refreshed from the new tail item via `parse_current_item`, with fallback to `lo_scanner.offset` when stack becomes empty), matching the tail-refresh semantics from `trim_back_to_upper_bound()`.

File: src/table/block/decoder.rs [hunk 2]
```diff
@@ -341,22 +392,36 @@ impl<'a, Item: Decodable<Parsed>, Parsed: ParsedItem<Item>> Decoder<'a, Item, Pa
...
-            Item::parse_full(reader, offset)
+            Item::parse_full(reader, offset, entries_end)
         } else {
             #[expect(clippy::expect_used, reason = "we trust the is_restart flag")]
             Item::parse_truncated(
                 reader,
                 offset,
                 base_key_offset.expect("should parse truncated item"),
+                base_key_end.expect("should parse truncated item"),
+                entries_end,
```
[the remark this answers is not in the document]
Comment (polaz) [reply src/table/block/decoder.rs]: @coderabbitai Fixed. `fill_stack()` now exits immediately when restart-head parse fails, and `Decoder::next()` only mutates interval state on successful parse; on parse failure it marks the forward scanner exhausted (`offset = block_end`, `remaining_in_interval = 0`). Added regression coverage ensuring repeated `next()` after malformed restart head stays `None` (no panic).

File: src/table/writer/mod.rs [hunk 3]
```diff
@@ -213,19 +213,30 @@ impl Writer {
...
 
     #[must_use]
     pub fn use_index_block_restart_interval(mut self, interval: u8) -> Self {
+        assert!(
+            interval > 0,
+            "index block restart interval must be greater than zero",
+        );
         self.index_block_restart_interval = interval;
+        self.index_writer = self.index_writer.use_restart_interval(interval);
         self
```
[the remark this answers is not in the document]
Comment (polaz) [reply src/table/writer/mod.rs]: @coderabbitai Fixed. Added lifecycle guards in both restart-interval setters to reject mutation after writing starts (`meta.data_block_count == 0 && chunk.is_empty()`), and added regression tests proving both setters panic after `write()`.

**Comment by polaz** [comment]
@coderabbitai re-review

**Comment by polaz** [comment]
@coderabbitai re-review

**Comment by polaz** [comment]
@coderabbitai Re: outside-diff "Binary search trusts semantically-corrupted binary index" (decoder.rs:245-299):

Out of scope for this perf PR. The scenario requires a corrupted binary index that still passes the block checksum (either encoder bug or checksum collision). Adding O(n) monotonicity validation to every seek call would regress the hot path for a case that checksums already catch. Tracked in #184 (fallible decoder constructor) which is the right place for comprehensive corruption hardening.
